src/ast.rs

```diff
@@ -44,240 +49,346 @@ lazy_static! {
+  ) -> Option<ParsedSource>;
+  fn get_parsed_source(
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Option<ParsedSource>;
 }
 
-/// Searches comments for any `@deno-types` compiler hints.
-pub fn analyze_deno_types(
-  desc: &DependencyDescriptor,
-) -> Option<(String, SourceRange)> {
-  let comment = desc.leading_comments.last()?;
-  let captures = DENO_TYPES_RE.captures(&comment.text)?;
-  if let Some(m) = captures.get(1) {
-    Some((
-      m.as_str().to_string(),
-      comment_match_to_source_range(comment, &m),
-    ))
-  } else if let Some(m) = captures.get(2) {
-    Some((
-      m.as_str().to_string(),
-      comment_match_to_source_range(comment, &m),
-    ))
-  } else {
-    unreachable!("Unexpected captures from deno types regex")
-  }
+/// Default store that works on a single thread.
+#[derive(Default)]
+pub struct DefaultParsedSourceStore {
+  store: RefCell<HashMap<ModuleSpecifier, ParsedSource>>,
 }
 
-/// Searches JSDoc comment blocks for type imports
-/// (e.g. `{import("./types.d.ts").Type}`) and returns a vector of tuples of
-/// the specifier and the span of the import.
-pub fn analyze_jsdoc_imports(
-  parsed_source: &ParsedSource,
-) -> Vec<(String, SourceRange)> {
-  let mut deps = Vec::new();
-  for comment in parsed_source.comments().get_vec().iter() {
-    if comment.kind != CommentKind::Block || !comment.text.starts_with('*') {
-      continue;
-    }
-    for captures in JSDOC_IMPORT_RE.captures_iter(&comment.text) {
-      if let Some(m) = captures.get(1) {
-        deps.push((
-          m.as_str().to_string(),
-          comment_match_to_source_range(comment, &m),
-        ));
-      }
-    }
+impl ParsedSourceStore for DefaultParsedSourceStore {
+  fn set_parsed_source(
+    &self,
+    specifier: ModuleSpecifier,
+    parsed_source: ParsedSource,
+  ) -> Option<ParsedSource> {
+    self.store.borrow_mut().insert(specifier, parsed_source)
+  }
+
+  fn get_parsed_source(
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Option<ParsedSource> {
+    self.store.borrow().get(specifier).cloned()
   }
-  deps
 }
 
-/// Searches comments for a `@jsxImportSource` pragma on JSX/TSX media types
-pub fn analyze_jsx_import_sources(
-  parsed_source: &ParsedSource,
-) -> Option<(String, SourceRange)> {
-  match parsed_source.media_type() {
-    MediaType::Jsx | MediaType::Tsx => {
-      parsed_source.get_leading_comments().iter().find_map(|c| {
-        let captures = JSX_IMPORT_SOURCE_RE.captures(&c.text)?;
-        let m = captures.get(1)?;
-        Some((m.as_str().to_string(), comment_match_to_source_range(c, &m)))
-      })
+/// Stores parsed files in the provided store after parsing.
+/// in a provided store.
+///
+/// Note that this will insert into the store whatever was
+/// last parsed, so if two threads race to parse, when they're
+/// both done it will have whatever was last stored.
+pub struct CapturingModuleParser {
+  parser: Box<dyn ModuleParser>,
+  store: Box<dyn ParsedSourceStore>,
```

Comment:
I've found these `Box`s severely limiting, so I'm going to replace them with references.